### ISSUE TITLE
Airplay to HomePod and Apple TV 4 disconnects randomly

### ISSUE THREAD

**Issue by retrography** [ejurgensen/forked-daapd#734]
I am running 26.5 on Ubuntu server and I have a very similar issue as described in #521 with my HomePod. The player simply stops streaming to my HomePod at some point in time. I don't see much of a regularity in timing apart from the fact that the playback rarely survives the second minute!

[Here](https://gist.github.com/retrography/1dc0f3a61af748d6e6876a6f7b236918) is the log.

**Comment by ejurgensen** [comment]
Do you always see these retransmit requests before a disconnect? It would appear that there are issues with packet loss on your network, and that it causes the HomePod to disconnect. I suggest you check if the issue persists on another network.

Also, please edit your post and add the log as an attachment instead of pasting the entire thing.

**Comment by retrography** [comment]
Interesting. I was actually wondering what those requests are. I will move the HomePod to another router and retry. I will also did the log above later today.

**Comment by retrography** [comment]
I am reopening this because I have tested all the conditions and fixed the packet loss issue to no avail. [Here](https://gist.github.com/retrography/7d56027640588548fcdcda7f5c8a48dc) is a new log. 

It is really strange how HomePod terminates the AirPlay connection. Any idea what is going wrong?

**Comment by ejurgensen** [comment]
Ok, let's reopen it. There are no clues in the log, and I see the hangup happens very quickly, so some keep-alive or sync issue doesn't seem likely. I will see if I can get my hands on a Homepod to reproduce.

**Comment by retrography** [comment]
If I increase the log level to spam, do you think we may get any clue?

More info: On the same network I have no problem streaming to shairport, and no problem streaming from Roon or Apple devices to HomePod. 

I do have a second router but it operates in access point mode to extend the range. I also tried turning it off earlier this morning and did not change the issue.

**Comment by dpisano** [comment]
I have starting having this problem too. For me it started after my HomePod was upgraded to 13.2. When it happens this is what I see in the logs:
```
[2019-11-11 04:45:29] [  LOG]     raop: Device 'Living Room' closed RTSP connection
[2019-11-11 04:45:29] [  LOG]   player: The AirPlay device 'Living Room' FAILED
```

**Comment by ejurgensen** [comment]
The chance that spam log level will produce clues is low, but you can always try. The upgrade that dpisano mentions sounds promising. Other than that I don’t know what you can do.

**Comment by dpisano** [comment]
I has able to get some more with debugging turned on. Not sure if it helps. Also it looks to be different messages now. 
```
[2019-11-13 03:35:30] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Living Room'
[2019-11-13 03:36:00] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Living Room'
[2019-11-13 03:36:30] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Living Room'
[2019-11-13 03:36:30] [  LOG]     raop: Keep alive request to 'Living Room' failed: 0 (null)
[2019-11-13 03:36:30] [  LOG]     raop: Device 'Living Room' closed RTSP connection
[2019-11-13 03:36:30] [DEBUG]     raop: Cleaning up failed session (deferred) on device 'Living Room'
[2019-11-13 03:36:30] [DEBUG]   player: Callback request received, id is 0
[2019-11-13 03:36:30] [DEBUG]   player: Unsubscription request for quality 44100/16/2 (now 0 subscribers)
[2019-11-13 03:36:30] [DEBUG]   player: Making deferred callback to device_streaming_cb, id was 0
[2019-11-13 03:36:30] [DEBUG]   player: Callback from AirPlay to device_streaming_cb (status -1)
[2019-11-13 03:36:30] [  LOG]   player: The AirPlay device 'Living Room' FAILED
[2019-11-13 03:36:30] [DEBUG]      mpd: Asynchronous listener callback called with event type 1.
[2019-11-13 03:36:30] [DEBUG]       db: Running query 'BEGIN TRANSACTION;'
[2019-11-13 03:36:30] [DEBUG]      mpd: Notify clients waiting for idle results: 1
[2019-11-13 03:36:30] [DEBUG]       db: Running query 'SELECT value FROM admin a WHERE a.key = 'queue_version';'
[2019-11-13 03:36:30] [DEBUG]       db: Running query 'DELETE FROM queue where id <> 0;'
[2019-11-13 03:36:30] [DEBUG]       db: Running query 'INSERT OR REPLACE INTO admin (key, value) VALUES ('queue_version', '11898');'
[2019-11-13 03:36:30] [DEBUG]       db: Running query 'END TRANSACTION;'
[2019-11-13 03:36:30] [ SPAM]   ffmpeg: Statistics: 35045376 bytes read, 0 seeks
[2019-11-13 03:36:30] [DEBUG]      mpd: Asynchronous listener callback called with event type 2.
[2019-11-13 03:36:30] [DEBUG]      mpd: Notify clients waiting for idle results: 2
[2019-11-13 03:36:33] [DEBUG]     mdns: Avahi Resolver: resolved service 'D4909CDB1891@Living Room' type '_raop._tcp' proto 0, host Living-Room.local, address 10.2.1.25
[2019-11-13 03:36:33] [DEBUG]     mdns: Connection test to 10.2.1.25:7000 completed successfully
[2019-11-13 03:36:33] [DEBUG]     raop: Event for AirPlay device 'Living Room' (port 7000, id d4909cdb1891)
[2019-11-13 03:36:33] [ INFO]     raop: Adding AirPlay device 'Living Room': password: 0, verification: 0, encrypt: 0, authsetup: 0, metadata: 1, type Other, address 10.2.1.25:7000
[2019-11-13 03:36:33] [DEBUG]      mpd: Asynchronous listener callback called with event type 8.
[2019-11-13 03:36:33] [DEBUG]      mpd: Notify clients waiting for idle results: 8
```

**Comment by retrography** [comment]
Further test: On the same network I have no issue streaming to AppleTV with FDAAPD.

**Comment by retrography** [comment]
I recompiled FDAAPD today and this problem seems to have been resolved. Configure failed on websockets and had me reinstall them. But apart from that I didn’t change anything. For an unknown reason it just works now.

**Comment by dpisano** [comment]
I'm still seeing the problem on my setup.

**Comment by holly79** [comment]
I'm still seeing the problem as well. I used the current release (27.0)

**Comment by ejurgensen** [comment]
I’ll reopen, and also try to borrow a Homepod to see if I can figure something out

**Comment by raufis27** [comment]
I have similar problem with HomePod. Just bought it. Every 10-15 min it will stop playing:

`[2020-01-12 23:23:23] [  LOG]     raop: Device 'LRPod' closed RTSP connection
[2020-01-12 23:23:23] [  LOG]   player: The AirPlay device 'LRPod' FAILED
[2020-01-12 23:36:21] [  LOG]     mdns: Avahi Resolver failure: service 'D4909CDF6D6F@LRPod' type '_raop._tcp' proto 1: Timeout reached
[2020-01-12 23:36:21] [  LOG]     mdns: Avahi Resolver failure: service 'D4909CDF6D6F@LRPod' type '_raop._tcp' proto 0: Timeout reached
[2020-01-12 23:36:51] [  LOG]   player: Output device with id 233717572267375 has disappeared from our list
[2020-01-12 23:37:10] [  LOG]   player: Output device with id 233717572267375 has disappeared from our list
[2020-01-12 23:37:10] [  LOG]   player: Output device with id 233717572267375 has disappeared from our list
[2020-01-12 23:38:52] [  LOG]   player: Output device with id 233717572267375 has disappeared from our list
[2020-01-12 23:48:56] [  LOG]     raop: Keep alive request to 'LRPod' failed: 0 (null)
[2020-01-12 23:48:56] [  LOG]     raop: Device 'LRPod' closed RTSP connection
[2020-01-12 23:48:56] [  LOG]   player: The AirPlay device 'LRPod' FAILED`

**Comment by raufis27** [comment]
I used this branch https://github.com/ejurgensen/forked-daapd/issues/855#issuecomment-570332501 to test and for past 30 min it is being stable.

**Comment by ejurgensen** [comment]
That’s a very nice offer, but I’m in Copenhagen, Denmark, and I see you’re in the US.

I haven’t gotten hold of a HomePod, but during next week I will be having a ATV 4k over. Maybe the error will reproduce on that.

**Comment by mixographer** [comment]
I have an interesting observation, and I'd like to ask the others what they are playing when the HomePod gives up. @raufis27 , @holly79 , @dpisano ? 

My HomePod seems stable when I play an AAC file at 320 VBR. It crashes very frequently when I play music from my Turntable. The Turntable creates a pipe with cpiped at 48000/16. 

I will now test creating my pipe at 44100/16 and see if that changes anything. 

My Airport express devices are stable with the 48000/16.

**Comment by ejurgensen** [comment]
I have tested for some hours with the ATV 4k, and the error did appear, but only once. So I wasn't able to learn much about it. I tried various ways I could think of to provoke the error, because without a way to reproduce it won't be possible to fix. I changed source quality, like you mention @mixographer, changed tracks, start/stop/pause, and also threw away large numbers of packets to simulate network issues. All to no avail. I also saw that the ATV does sync clock, so that part seems ok.

The error appeared relatively early, after something like 10 minutes, and I did note some weird behaviour leading up to it:
- the ATV was making strange retransmit requests, at one instance asking for packets that were 12 seconds older than what forked-daapd was currently sending
- the ATV had brief playback interruptions, like it was underrunning

Right before the disconnect the ATV also made unusual retransmit requests. I was on purpose discarding every 1000th packet, so the first two retransmit requests are expected, but the third is peculiar, since it is 3 seconds late:
```
[2020-01-17 20:05:12] [ WARN]     raop: Discarding seqnum 5000
[2020-01-17 20:05:12] [DEBUG]     raop: Got retransmit request from 'Living Room': seqnum 5000 (len 1), last RTP session seqnum 5002 (len 1000)

[2020-01-17 20:05:20] [ WARN]     raop: Discarding seqnum 6000
[2020-01-17 20:05:20] [DEBUG]     raop: Got retransmit request from 'Living Room': seqnum 6000 (len 1), last RTP session seqnum 6001 (len 1000)

[2020-01-17 20:05:28] [ WARN]     raop: Discarding seqnum 7000
[2020-01-17 20:05:31] [DEBUG]     raop: Got retransmit request from 'Living Room': seqnum 7000 (len 1), last RTP session seqnum 7438 (len 1000)

[2020-01-17 20:05:32] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Living Room'
[2020-01-17 20:05:32] [  LOG]     raop: Keep alive request to 'Living Room' failed: 0 (null)
```

To be honest, I'm not sure what to make of this.

**Comment by mixographer** [comment]
I was going to go down some of the paths you mentioned in the AppleTV issue. It looks like you've started down that track. I'll go after those tomorrow. 

Why did it discard even 1000 seqnums? 
(Discarding seqnum 6000, 7000 etc?)

I got home from work, and I've been playing some records for a few hours tonight with no issues. I did connect my server with an ethernet cable since I last saw a HomePod fail.

**Comment by mixographer** [comment]
If I want to test my changes, what's the best way to re-compile and just install the changes? I see that `make install-exec` overwrites my forked-daapd.conf, so I don't think I want to keep doing that.

**Comment by ejurgensen** [comment]
One option is to not use `make install`, but instead just `make` and then run `~/builddir/src/forked-daapd -f`. This also means that /usr/sbin/forked-daapd will not get overwritten, which is good if you are testing on the same machine where you have forked-daapd installed with a packet manager.

**Comment by holly79** [comment]
> I have an interesting observation, and I'd like to ask the others what they are playing when the HomePod gives up. @raufis27 , @holly79 , @dpisano ?
> 
> My HomePod seems stable when I play an AAC file at 320 VBR. It crashes very frequently when I play music from my Turntable. The Turntable creates a pipe with cpiped at 48000/16.
> 
> I will now test creating my pipe at 44100/16 and see if that changes anything.
> 
> My Airport express devices are stable with the 48000/16.

I used forked-daapd with internet radio streams, so I haven't any AAC files at 320 VBR. At the first time I thought I have some trouble with my internet connection, but if I streamed internet radio to Homepods and shareport-sync client at the same time, only the Homepods was stopping.

**Comment by holly79** [comment]
Here you can find the sequence between the latest successfully keep_alive (17:16:11) and failed kepp_alive (17:17:11) of my Homepod "Küche". The other speaker are Homepods as well. May be it helps to solve the issue.

[2020-01-19 17:16:11] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Badezimmer'
[2020-01-19 17:16:11] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Esszimmer'
[2020-01-19 17:16:11] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Küche'
[2020-01-19 17:16:12] [ SPAM]   laudio: counter 60/100, stamp 625482:9, now 625743:15, elapsed is 261006 ms, cur_pos=11510323, exp_pos=11510364, diff=-41
[2020-01-19 17:16:13] [ SPAM]   laudio: counter 61/100, stamp 625482:9, now 625744:14, elapsed is 262005 ms, cur_pos=11554404, exp_pos=11554420, diff=-16
[2020-01-19 17:16:14] [ SPAM]   laudio: counter 62/100, stamp 625482:9, now 625745:14, elapsed is 263005 ms, cur_pos=11598499, exp_pos=11598520, diff=-21
[2020-01-19 17:16:15] [ SPAM]   laudio: counter 63/100, stamp 625482:9, now 625746:15, elapsed is 264006 ms, cur_pos=11642622, exp_pos=11642664, diff=-42
[2020-01-19 17:16:16] [ SPAM]   laudio: counter 64/100, stamp 625482:9, now 625747:14, elapsed is 265005 ms, cur_pos=11686702, exp_pos=11686720, diff=-18
[2020-01-19 17:16:17] [ SPAM]   laudio: counter 65/100, stamp 625482:9, now 625748:15, elapsed is 266005 ms, cur_pos=11730776, exp_pos=11730820, diff=-44
[2020-01-19 17:16:18] [ SPAM]   laudio: counter 66/100, stamp 625482:9, now 625749:15, elapsed is 267006 ms, cur_pos=11774924, exp_pos=11774964, diff=-40
[2020-01-19 17:16:19] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 50631 (len 1), last RTP session seqnum 50634 (len 1000)
[2020-01-19 17:16:19] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 50632 (len 1), last RTP session seqnum 50634 (len 1000)
[2020-01-19 17:16:19] [ SPAM]   laudio: counter 67/100, stamp 625482:9, now 625750:15, elapsed is 268006 ms, cur_pos=11819031, exp_pos=11819064, diff=-33
[2020-01-19 17:16:20] [ SPAM]   laudio: counter 68/100, stamp 625482:9, now 625751:15, elapsed is 269005 ms, cur_pos=11863122, exp_pos=11863120, diff=2
[2020-01-19 17:16:21] [ SPAM]   laudio: counter 69/100, stamp 625482:9, now 625752:14, elapsed is 270005 ms, cur_pos=11907204, exp_pos=11907220, diff=-16
[2020-01-19 17:16:22] [ SPAM]   laudio: counter 70/100, stamp 625482:9, now 625753:15, elapsed is 271005 ms, cur_pos=11951323, exp_pos=11951320, diff=3
[2020-01-19 17:16:23] [ SPAM]   laudio: counter 71/100, stamp 625482:9, now 625754:15, elapsed is 272006 ms, cur_pos=11995431, exp_pos=11995464, diff=-33
[2020-01-19 17:16:24] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 51307 (len 1), last RTP session seqnum 51310 (len 1000)
[2020-01-19 17:16:24] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 51308 (len 1), last RTP session seqnum 51310 (len 1000)
[2020-01-19 17:16:24] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 51309 (len 1), last RTP session seqnum 51310 (len 1000)
[2020-01-19 17:16:24] [ SPAM]   laudio: counter 72/100, stamp 625482:9, now 625755:15, elapsed is 273006 ms, cur_pos=12039540, exp_pos=12039564, diff=-24
[2020-01-19 17:16:25] [ SPAM]   laudio: counter 73/100, stamp 625482:9, now 625756:15, elapsed is 274006 ms, cur_pos=12083614, exp_pos=12083664, diff=-50
[2020-01-19 17:16:26] [ SPAM]   laudio: counter 74/100, stamp 625482:9, now 625757:18, elapsed is 275009 ms, cur_pos=12127894, exp_pos=12127896, diff=-2
[2020-01-19 17:16:27] [ SPAM]   laudio: counter 75/100, stamp 625482:9, now 625758:15, elapsed is 276005 ms, cur_pos=12171830, exp_pos=12171820, diff=10
[2020-01-19 17:16:28] [ SPAM]   laudio: counter 76/100, stamp 625482:9, now 625759:15, elapsed is 277005 ms, cur_pos=12215895, exp_pos=12215920, diff=-25
[2020-01-19 17:16:29] [ SPAM]   laudio: counter 77/100, stamp 625482:9, now 625760:15, elapsed is 278005 ms, cur_pos=12260028, exp_pos=12260020, diff=8
[2020-01-19 17:16:30] [ SPAM]   laudio: counter 78/100, stamp 625482:9, now 625761:15, elapsed is 279006 ms, cur_pos=12304140, exp_pos=12304164, diff=-24
[2020-01-19 17:16:31] [ SPAM]   laudio: counter 79/100, stamp 625482:9, now 625762:14, elapsed is 280005 ms, cur_pos=12348212, exp_pos=12348220, diff=-8
[2020-01-19 17:16:32] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 52332 (len 1), last RTP session seqnum 52337 (len 1000)
[2020-01-19 17:16:32] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 52333 (len 1), last RTP session seqnum 52337 (len 1000)
[2020-01-19 17:16:32] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 52334 (len 1), last RTP session seqnum 52337 (len 1000)
[2020-01-19 17:16:32] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 52335 (len 1), last RTP session seqnum 52337 (len 1000)
[2020-01-19 17:16:32] [ SPAM]   laudio: counter 80/100, stamp 625482:9, now 625763:15, elapsed is 281006 ms, cur_pos=12392348, exp_pos=12392364, diff=-16
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 52394 (len 1), last RTP session seqnum 52404 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 52395 (len 1), last RTP session seqnum 52404 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 52396 (len 1), last RTP session seqnum 52404 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 52397 (len 1), last RTP session seqnum 52404 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 52398 (len 1), last RTP session seqnum 52404 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 52399 (len 1), last RTP session seqnum 52404 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 52400 (len 1), last RTP session seqnum 52404 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52432 (len 1), last RTP session seqnum 52448 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52433 (len 1), last RTP session seqnum 52448 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52434 (len 1), last RTP session seqnum 52448 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52435 (len 1), last RTP session seqnum 52448 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52436 (len 1), last RTP session seqnum 52448 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52437 (len 1), last RTP session seqnum 52448 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52438 (len 1), last RTP session seqnum 52449 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52439 (len 1), last RTP session seqnum 52449 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52440 (len 1), last RTP session seqnum 52449 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52441 (len 1), last RTP session seqnum 52449 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52442 (len 1), last RTP session seqnum 52449 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52443 (len 1), last RTP session seqnum 52449 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 52444 (len 1), last RTP session seqnum 52449 (len 1000)
[2020-01-19 17:16:33] [ SPAM]   laudio: counter 81/100, stamp 625482:9, now 625764:15, elapsed is 282006 ms, cur_pos=12436447, exp_pos=12436464, diff=-17
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 52501 (len 1), last RTP session seqnum 52504 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 52502 (len 1), last RTP session seqnum 52504 (len 1000)
[2020-01-19 17:16:33] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 52503 (len 1), last RTP session seqnum 52504 (len 1000)
[2020-01-19 17:16:34] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 52540 (len 1), last RTP session seqnum 52543 (len 1000)
[2020-01-19 17:16:34] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 52541 (len 1), last RTP session seqnum 52543 (len 1000)
[2020-01-19 17:16:34] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 52542 (len 1), last RTP session seqnum 52543 (len 1000)
[2020-01-19 17:16:34] [ SPAM]   laudio: counter 82/100, stamp 625482:9, now 625765:23, elapsed is 283013 ms, cur_pos=12480882, exp_pos=12480873, diff=9
[2020-01-19 17:16:35] [ SPAM]   laudio: counter 83/100, stamp 625482:9, now 625766:16, elapsed is 284007 ms, cur_pos=12524703, exp_pos=12524708, diff=-5
[2020-01-19 17:16:36] [ SPAM]   laudio: counter 84/100, stamp 625482:9, now 625767:14, elapsed is 285005 ms, cur_pos=12568716, exp_pos=12568720, diff=-4
[2020-01-19 17:16:37] [ SPAM]   laudio: counter 85/100, stamp 625482:9, now 625768:15, elapsed is 286006 ms, cur_pos=12612851, exp_pos=12612864, diff=-13
[2020-01-19 17:16:38] [ SPAM]   laudio: counter 86/100, stamp 625482:9, now 625769:15, elapsed is 287005 ms, cur_pos=12656936, exp_pos=12656920, diff=16
[2020-01-19 17:16:39] [ SPAM]   laudio: counter 87/100, stamp 625482:9, now 625770:14, elapsed is 288005 ms, cur_pos=12701003, exp_pos=12701020, diff=-17
[2020-01-19 17:16:40] [ SPAM]   laudio: counter 88/100, stamp 625482:9, now 625771:15, elapsed is 289005 ms, cur_pos=12745137, exp_pos=12745120, diff=17
[2020-01-19 17:16:41] [ SPAM]   laudio: counter 89/100, stamp 625482:9, now 625772:15, elapsed is 290006 ms, cur_pos=12789242, exp_pos=12789264, diff=-22
[2020-01-19 17:16:41] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Badezimmer'
[2020-01-19 17:16:41] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Esszimmer'
[2020-01-19 17:16:41] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Küche'
[2020-01-19 17:16:42] [ SPAM]   laudio: counter 90/100, stamp 625482:9, now 625773:16, elapsed is 291007 ms, cur_pos=12833371, exp_pos=12833408, diff=-37
[2020-01-19 17:16:43] [ SPAM]   laudio: counter 91/100, stamp 625482:9, now 625774:15, elapsed is 292006 ms, cur_pos=12877444, exp_pos=12877464, diff=-20
[2020-01-19 17:16:44] [ SPAM]   laudio: counter 92/100, stamp 625482:9, now 625775:15, elapsed is 293006 ms, cur_pos=12921558, exp_pos=12921564, diff=-6
[2020-01-19 17:16:45] [ SPAM]   laudio: counter 93/100, stamp 625482:9, now 625776:16, elapsed is 294007 ms, cur_pos=12965720, exp_pos=12965708, diff=12
[2020-01-19 17:16:46] [ SPAM]   laudio: counter 94/100, stamp 625482:9, now 625777:14, elapsed is 295005 ms, cur_pos=13009729, exp_pos=13009720, diff=9
[2020-01-19 17:16:47] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 54228 (len 1), last RTP session seqnum 54230 (len 1000)
[2020-01-19 17:16:47] [DEBUG]     raop: Got retransmit request from 'Küche': seqnum 54229 (len 1), last RTP session seqnum 54230 (len 1000)
[2020-01-19 17:16:47] [ SPAM]   laudio: counter 95/100, stamp 625482:9, now 625778:15, elapsed is 296005 ms, cur_pos=13053843, exp_pos=13053820, diff=23
[2020-01-19 17:16:48] [ SPAM]   laudio: counter 96/100, stamp 625482:9, now 625779:15, elapsed is 297006 ms, cur_pos=13097950, exp_pos=13097964, diff=-14
[2020-01-19 17:16:49] [ SPAM]   laudio: counter 97/100, stamp 625482:9, now 625780:15, elapsed is 298006 ms, cur_pos=13142049, exp_pos=13142064, diff=-15
[2020-01-19 17:16:50] [ SPAM]   laudio: counter 98/100, stamp 625482:9, now 625781:15, elapsed is 299006 ms, cur_pos=13186136, exp_pos=13186164, diff=-28
[2020-01-19 17:16:51] [ SPAM]   laudio: counter 99/100, stamp 625482:9, now 625782:14, elapsed is 300005 ms, cur_pos=13230228, exp_pos=13230220, diff=8
[2020-01-19 17:16:51] [DEBUG]   laudio: Sync check result: drift=1.389451, latency=-61.865275, r2=0.774132, sync=0
[2020-01-19 17:16:52] [ SPAM]   laudio: counter 0/100, stamp 625482:9, now 625783:15, elapsed is 301006 ms, cur_pos=13274352, exp_pos=13274364, diff=-12
[2020-01-19 17:16:53] [ SPAM]   laudio: counter 1/100, stamp 625482:9, now 625784:14, elapsed is 302005 ms, cur_pos=13318441, exp_pos=13318420, diff=21
[2020-01-19 17:16:54] [ SPAM]   laudio: counter 2/100, stamp 625482:9, now 625785:14, elapsed is 303005 ms, cur_pos=13362529, exp_pos=13362520, diff=9
[2020-01-19 17:16:55] [ SPAM]   laudio: counter 3/100, stamp 625482:9, now 625786:14, elapsed is 304005 ms, cur_pos=13406632, exp_pos=13406620, diff=12
[2020-01-19 17:16:55] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 55260 (len 1), last RTP session seqnum 55262 (len 1000)
[2020-01-19 17:16:56] [ SPAM]   laudio: counter 4/100, stamp 625482:9, now 625787:16, elapsed is 305006 ms, cur_pos=13450797, exp_pos=13450764, diff=33
[2020-01-19 17:16:57] [ SPAM]   laudio: counter 5/100, stamp 625482:9, now 625788:15, elapsed is 306006 ms, cur_pos=13494857, exp_pos=13494864, diff=-7
[2020-01-19 17:16:58] [ SPAM]   laudio: counter 6/100, stamp 625482:9, now 625789:15, elapsed is 307006 ms, cur_pos=13538960, exp_pos=13538964, diff=-4
[2020-01-19 17:16:59] [ SPAM]   laudio: counter 7/100, stamp 625482:9, now 625790:15, elapsed is 308006 ms, cur_pos=13583053, exp_pos=13583064, diff=-11
[2020-01-19 17:17:00] [ SPAM]   laudio: counter 8/100, stamp 625482:9, now 625791:17, elapsed is 309007 ms, cur_pos=13627240, exp_pos=13627208, diff=32
[2020-01-19 17:17:01] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 55896 (len 1), last RTP session seqnum 55899 (len 1000)
[2020-01-19 17:17:01] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 55897 (len 1), last RTP session seqnum 55899 (len 1000)
[2020-01-19 17:17:01] [DEBUG]     raop: Got retransmit request from 'Badezimmer': seqnum 55898 (len 1), last RTP session seqnum 55899 (len 1000)
[2020-01-19 17:17:01] [ SPAM]   laudio: counter 9/100, stamp 625482:9, now 625792:15, elapsed is 310005 ms, cur_pos=13671247, exp_pos=13671220, diff=27
[2020-01-19 17:17:02] [ SPAM]   laudio: counter 10/100, stamp 625482:9, now 625793:15, elapsed is 311006 ms, cur_pos=13715358, exp_pos=13715364, diff=-6
[2020-01-19 17:17:03] [ SPAM]   laudio: counter 11/100, stamp 625482:9, now 625794:15, elapsed is 312005 ms, cur_pos=13759454, exp_pos=13759420, diff=34
[2020-01-19 17:17:04] [ SPAM]   laudio: counter 12/100, stamp 625482:9, now 625795:15, elapsed is 313006 ms, cur_pos=13803538, exp_pos=13803564, diff=-26
[2020-01-19 17:17:05] [ SPAM]   laudio: counter 13/100, stamp 625482:9, now 625796:14, elapsed is 314005 ms, cur_pos=13847640, exp_pos=13847620, diff=20
[2020-01-19 17:17:06] [ SPAM]   laudio: counter 14/100, stamp 625482:9, now 625797:14, elapsed is 315005 ms, cur_pos=13891736, exp_pos=13891720, diff=16
[2020-01-19 17:17:07] [ SPAM]   laudio: counter 15/100, stamp 625482:9, now 625798:15, elapsed is 316006 ms, cur_pos=13935823, exp_pos=13935864, diff=-41
[2020-01-19 17:17:08] [ SPAM]   laudio: counter 16/100, stamp 625482:9, now 625799:14, elapsed is 317005 ms, cur_pos=13979938, exp_pos=13979920, diff=18
[2020-01-19 17:17:09] [ SPAM]   laudio: counter 17/100, stamp 625482:9, now 625800:15, elapsed is 318006 ms, cur_pos=14024066, exp_pos=14024064, diff=2
[2020-01-19 17:17:10] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 57030 (len 1), last RTP session seqnum 57033 (len 1000)
[2020-01-19 17:17:10] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 57031 (len 1), last RTP session seqnum 57033 (len 1000)
[2020-01-19 17:17:10] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 57032 (len 1), last RTP session seqnum 57033 (len 1000)
[2020-01-19 17:17:10] [ SPAM]   laudio: counter 18/100, stamp 625482:9, now 625801:14, elapsed is 319005 ms, cur_pos=14068154, exp_pos=14068120, diff=34
[2020-01-19 17:17:11] [ SPAM]   laudio: counter 19/100, stamp 625482:9, now 625802:14, elapsed is 320005 ms, cur_pos=14112256, exp_pos=14112220, diff=36
[2020-01-19 17:17:11] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Badezimmer'
[2020-01-19 17:17:11] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Esszimmer'
[2020-01-19 17:17:11] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Küche'
[2020-01-19 17:17:11] [  LOG]     raop: Keep alive request to 'Küche' failed: 0 (null)
[2020-01-19 17:17:11] [  LOG]     raop: Device 'Küche' closed RTSP connection
[2020-01-19 17:17:11] [DEBUG]     raop: Cleaning up failed session (deferred) on device 'Küche'
[2020-01-19 17:17:11] [DEBUG]   player: Callback request received, id is 0
[2020-01-19 17:17:11] [DEBUG]   player: Making deferred callback to device_streaming_cb, id was 0
[2020-01-19 17:17:11] [DEBUG]   player: Callback from AirPlay to device_streaming_cb (status -1)
[2020-01-19 17:17:11] [  LOG]   player: The AirPlay device 'Küche' FAILED
[2020-01-19 17:17:11] [DEBUG]   player: 1. Active callback: device_streaming_cb
[2020-01-19 17:17:11] [DEBUG]   player: 2. Active callback: device_streaming_cb
[2020-01-19 17:17:11] [DEBUG]   player: 3. Active callback: device_streaming_cb
[2020-01-19 17:17:12] [ SPAM]   laudio: counter 20/100, stamp 625482:9, now 625803:15, elapsed is 321006 ms, cur_pos=14156365, exp_pos=14156364, diff=1
[2020-01-19 17:17:12] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 57377 (len 1), last RTP session seqnum 57385 (len 1000)
[2020-01-19 17:17:12] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 57378 (len 1), last RTP session seqnum 57385 (len 1000)
[2020-01-19 17:17:12] [DEBUG]     raop: Got retransmit request from 'Esszimmer': seqnum 57379 (len 1), last RTP session seqnum 57385 (len 1000)
[2020-01-19 17:17:13] [ SPAM]   laudio: counter 21/100, stamp 625482:9, now 625804:14, elapsed is 322005 ms, cur_pos=14200441, exp_pos=14200420, diff=21
[2020-01-19 17:17:14] [DEBUG]     mdns: Avahi Resolver: resolved service '50BC96014E8D@Küche' type '_raop._tcp' proto 0, host Kuche.local, address 192.168.200.149
[2020-01-19 17:17:14] [DEBUG]     mdns: Connection test to 192.168.200.149:7000 completed successfully
[2020-01-19 17:17:14] [DEBUG]     raop: Event for AirPlay device 'Küche' (port 7000, id 50bc96014e8d)
[2020-01-19 17:17:14] [ INFO]     raop: Adding AirPlay device 'Küche': password: 0, verification: 0, encrypt: 0, authsetup: 0, metadata: 1, type Other, address 192.168.200.149:7000
[2020-01-19 17:17:14] [DEBUG]      mpd: Asynchronous listener callback called with event type 8.
[2020-01-19 17:17:14] [DEBUG]      mpd: Notify clients waiting for idle results: 8
[2020-01-19 17:17:14] [ SPAM]      web: LWS fd=93, revents=4
[2020-01-19 17:17:14] [ SPAM]      web: LWS lws_calllback_as_writeable: 0xa8a16880 (user=0xa8a04ac8)
[2020-01-19 17:17:14] [DEBUG]      web: notify callback reason: 11
[2020-01-19 17:17:14] [ SPAM]      web: LWS _lws_rx_flow_control: no pending change
[2020-01-19 17:17:14] [ SPAM]      web: LWS fd=47, revents=4
[2020-01-19 17:17:14] [ SPAM]      web: LWS lws_calllback_as_writeable: 0xa8a04e08 (user=0xa8a04a48)
[2020-01-19 17:17:14] [DEBUG]      web: notify callback reason: 11
[2020-01-19 17:17:14] [ SPAM]      web: LWS _lws_rx_flow_control: no pending change
[2020-01-19 17:17:14] [ SPAM]   laudio: counter 22/100, stamp 625482:9, now 625805:15, elapsed is 323005 ms, cur_pos=14244558, exp_pos=14244520, diff=38
[2020-01-19 17:17:15] [ SPAM]   laudio: counter 23/100, stamp 625482:9, now 625806:14, elapsed is 324005 ms, cur_pos=14288644, exp_pos=14288620, diff=24
[2020-01-19 17:17:16] [ SPAM]   laudio: counter 24/100, stamp 625482:9, now 625807:14, elapsed is 325005 ms, cur_pos=14332744, exp_pos=14332720, diff=24
[2020-01-19 17:17:17] [ SPAM]   laudio: counter 25/100, stamp 625482:9, now 625808:15, elapsed is 326006 ms, cur_pos=14376870, exp_pos=14376864, diff=6

**Comment by ejurgensen** [comment]
Please edit your post, large log snippets make the thread hard to read. If you don't know what is relevant to include, then attach the log as file.

**Comment by ejurgensen** [comment]
Thanks for the edit. Your log seems to show very normal playback with normal retransmit requests, and then a failure like observed above. I notice all the failures seem to happen when forked-daapd makes the keep_alive request (which is made each minute). Seems like it might be a clue? @mixographer if you are experimenting you could try increasing the keep_alive frequency to see if that provokes faster failure.

**Comment by dpisano** [comment]
I see failure ether within the first minute or many hours later. I'm also sending an internet radio stream to it.

With the keep_alive, how many in a row can fail?

**Comment by ejurgensen** [comment]
>With the keep_alive, how many in a row can fail?

Not sure I understand the question. It is not forked-daapd that is disconnecting, it is the Homepod. The reason for that may be related to the keep_alive's, but it is not because a certain number of them are failing. You can see from the logs that the keep_alive's are not failing until the disconnect.

**Comment by dpisano** [comment]
The question comes from my lack of understanding of the airplay protocol. I deal a lot with network protocols that allow for a certain number of keep alive packets to be missed before dropping the connection. So the question came from my understanding of those protocols.

**Comment by ejurgensen** [comment]
The keep-alive is more of a hack than part of the protocol. It solved issue #521 which was an issue that had a lot in common with this one. So maybe it has made a comeback.

**Comment by mixographer** [comment]
I was noticing that the failure appeared to come at the same time as the keep-alive, but is that when the HomePod disconnects? or is that the first time we realize that it didn't respond? I haven't set the homePod next to my pc so I can watch the log, and hear the music stop at the same time....I'll try that.

**Comment by ejurgensen** [comment]
That's a good question. Forked-daapd has a number of connections to the speaker, one is a tcp control connection. This is the connection that is used to send the keep_alives, and it is also the connection that gets the disconnect event which is logged. The audio is sent over an udp connection, and I can't see if that is also disconnected. So perhaps it might work to just reconnect the control instead of failing? As an experiment, you could also try disabling line 2700 and 2701 in raop.c and see if that makes a difference:
https://github.com/ejurgensen/forked-daapd/blob/master/src/outputs/raop.c#L2700
That should make forked-daapd ignore the disconnect, so then you can better see if the music continues.

**Comment by mixographer** [comment]
I guess I don't understand line 2701. 

 if (!rs->reqs_in_flight)
    evrtsp_connection_set_closecb(rs->ctrl, raop_rtsp_close_cb, rs);

Does this need any {} curly brackets to make it part of the conditional? If I change 
2701 to a log item instead of this closing the callback, it logs the event and the music keeps playing. and the HomePod is still responding to volume events if I send them. 

It seems like every time the options are sent, the result triggers ' if (!rs->reqs_in_flight)'

**Comment by mixographer** [comment]
Changed line 2701 in rap.c to print the log:

> if (!rs->reqs_in_flight)
>     {
>       DPRINTF(E_LOG, L_RAOP, "***** !rs->reqs_in_flight case: not calling evtrtsp_connection_set_closecb().\n");
>       //   evrtsp_connection_set_closecb(rs->ctrl, raop_rtsp_close_cb, rs);
>     }
> 

and this is the log with every keep alive:
[DEBUG]     raop: keep_alive: Sending OPTIONS to 'living room stereo'
[DEBUG]     raop: keep_alive: Sending OPTIONS to 'Bathroom'
[  LOG]     raop: ***** !rs->reqs_in_flight case: not calling evtrtsp_connection_set_closecb().
[  LOG]     raop: ***** !rs->reqs_in_flight case: not calling evtrtsp_connection_set_closecb().

**Comment by holly79** [comment]
I tried a version with commented line 2701, but my homepods stop after a while. But not all at the same time.

**Comment by ejurgensen** [comment]
To be clear, disabling those lines was only to test if the device would keep on playing for a while after the keep_alive. I did not expect it to fix anything.

If you are up to it, there is another test I would like you to make. To do that you need to build with the change above, which you can get if you checkout the [atv_disconnect1 ](https://github.com/ejurgensen/forked-daapd/tree/atv_disconnect1)branch. The change should make forked-daapd try to reconnect if the keep_alive fails. So please watch the log for lines with "RTSP reconnect ....", and especially if playback continues after such a line is logged.

**Comment by holly79** [comment]
> To be clear, disabling those lines was only to test if the device would keep on playing for a while after the keep_alive. I did not expect it to fix anything.
> 
for sure, it was only for testing

> If you are up to it, there is another test I would like you to make. To do that you need to build with the change above, which you can get if you checkout the [atv_disconnect1 ](https://github.com/ejurgensen/forked-daapd/tree/atv_disconnect1)branch. The change should make forked-daapd try to reconnect if the keep_alive fails. So please watch the log for lines with "RTSP reconnect ....", and especially if playback continues after such a line is logged.

I will try it and let you know the result.

**Comment by mixographer** [comment]
I haven't had a chance to checkout the latest branch, but I will. 

I will also say, that I made the following changes: 

- changed the packet buffer to 2000 (RAOP_PACKET_BUFFER_SIZE    2000) in raop.c 
- changed the time threshold to 3 seconds in config (cache_daap_threshold = 3000)
- disabled ipv6 in config (ipv6 = no) 
- disabled ipv6 system wide in ubuntu. 

I have not had any HomePod dropouts since. It could be coincidence.

**Comment by holly79** [comment]
I had a chance to compile and test the branch. But I have no good news. I saw the message "raop: RTSP reconnect to 'Esszimmer' succeeded", but the Homepod stopped the music. I will try the changes of mixographer.

**Comment by ejurgensen** [comment]
Thanks for testing. So we can conclude that reconnecting the control connection isn't enough.

Of the changes listed by mixographer I am pretty sure the time threshold has no impact. It is also unlikely that the packet buffer size has, I would say. ipv6 might make a difference, so maybe just try with that. But as mixographer says, it may just be coincidences.

**Comment by mixographer** [comment]
Looks like I have lost the ability to play to any HomePod now. I think Apple may have changed something with an update. My HomePods also randomly play songs and streams I don't ask them to play.

**Comment by skabber** [comment]
I just wanted to drop a note here that I am having the same issue. I use forked-daapd to stream my turn table to 3 HomePods and 1 Sonos One. The HomePods are always disconnecting and the Sonos is rock solid over AirPlay.

**Comment by ejurgensen** [comment]
Thanks for the note. Would you be able to make a test for me? I have made a change in the atv_disconnect2 branch, and I would like to know if it has any effect on the issue. The change is about using another type of request for "heartbeat". Here is a link to the branch: https://github.com/ejurgensen/forked-daapd/tree/atv_disconnect2

Let me know if you need help with building/testing.

**Comment by skabber** [comment]
Sure thing. I got it building but having it fails to launch because it can't open the sqlite db. I think this might because of the note about special building sqlite with the SQLITE_ENABLE_UNLOCK_NOTIFY enabled. Do you have any tips for that. I'm compiling on a RaspberryPi4.

**Comment by ejurgensen** [comment]
I think the reason is more simple, it is probable just the permissions. When you do "make install" a new config file gets installed, possibly overwriting your old one. So that might mean that either the uid or locations are no longer correct (see the "general" section of the config file).

**Comment by raufis27** [comment]
> Thanks for the note. Would you be able to make a test for me? I have made a change in the atv_disconnect2 branch, and I would like to know if it has any effect on the issue. The change is about using another type of request for "heartbeat". Here is a link to the branch: https://github.com/ejurgensen/forked-daapd/tree/atv_disconnect2
> 
> Let me know if you need help with building/testing.

Same problem with this build :(

**Comment by skabber** [comment]
Yup. Same for me. I just had one of my home pods drop out. I’ll collect some debug logs and post them.

**Comment by ejurgensen** [comment]
Guess it was a dud, but thanks for testing both of you. Unfortunately, I don't think debug logs will help - there are already some of those above, and they just say that the device hangs up.

**Comment by retrography** [comment]
Is there any other open implementation of an AirPlay transmitter that we can use to investigate this issue? Just out of curiosity.

**Comment by ejurgensen** [comment]
Yes, might be worth trying. I haven't tried any alternatives myself, but perhaps these work?
https://github.com/philippe44/RAOP-Player
https://github.com/chevil/raop2_play

**Comment by retrography** [comment]
raop2_play seems to be a dead fork. The original RAOP-Player compiles with no problem, and plays to HomePod. Surprise, surprise, it comes down with the same error after a few seconds. Increasing the log level does not reveal anything. Even worse, RAOP-Player doesn't even notice its connection has been terminated. It keeps reporting back its playing status while HomePod is in idle mode.

What I find interesting is that the changes between v26.5 and v27.0 (of Forked-DAAPD) have made the connection much more stable for me. I still have the eventual dropout, but much less frequently, and the playback continues for longer timespans. FDAAPD used to behave like RAOP-Player before, losing connection within the first minute. But now it can stay connected for hours sometimes, although most of the time it loses connection within an hour or so.

**Comment by ejurgensen** [comment]
Ok thanks for testing this so quickly. When you tested with RAOP-Player did you then use any sync? I.e. the NTP options. I have been thinking if the issue relates to the timing, but if the issue is also present with RAOP-Player and no sync, then I guess that is unlikely.

Your observation is interesting. Between 26.5 and 27.0 there was a major refactor of the player, and timing is one of the things that was changed. But there were many other, so hard to pinpoint or test what might have made a difference.

**Comment by retrography** [comment]
Is there any specific settings you want me to dabble with? I tried setting NTP manually, to no avail.

Interestingly, I found a thread on HomePod issues with another RAOP projects developed by the same developer from back in 2018 that seems to have been resolved. They claim that the disconnection issue had to do with not transferring all the metadata that HomePod expected. See for yourself: https://github.com/philippe44/LMS-to-Raop/issues/4

**Comment by ejurgensen** [comment]
Interesting link. If this plugin is working it would be interesting to compare Wireshark traffic dumps - I don't know if you can produce those?

forked-daapd does send the metadata you mention, but who knows if there is some detail that is different and causes the disconnect. A dump might reveal that.

Another thing that you could test is if it matters how often the metadata is sent. I think forked-daapd only does it when the track changes, so you could set some short tracks to play on repeat and see if there is still a disconnect.

**Comment by holly79** [comment]
I tried atv_diconnect2 branch to check if it is working with short tracks (30-35sec). My Hompod plays over 2h. I check that with 1-2min tracks and it stopped working after few minutes.

**Comment by ejurgensen** [comment]
Sounds like a possible break-through! I'll make a branch asap that sends metadata every 30 sec, maybe that will do the trick.

**Comment by ejurgensen** [comment]
This metadata thing looks promising - I could now both reproduce the issue on my ATV4 by playing a very long track, and I could avoid any disconnect by sending progress metadata every 30 sec. If anyone wants to try it is in the new atv_disconnect3 branch.

Even though it seems to work, I think the progress metadata is incorrect when being sent continuously like this. So I need to work on those calculations.

**Comment by mixographer** [comment]
I've built this branch and I'll give it a try. I'll let you know what I find.

**Comment by mixographer** [comment]
It dropped again with this log, interesting that as soon as it drops, it tries and succeeds in resolving the host.

[2020-04-20 14:09:37] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Kitchen'
[2020-04-20 14:09:37] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Family Room'
[2020-04-20 14:09:37] [  LOG]     raop: ***** !rs->reqs_in_flight case: not calling evtrtsp_connection_set_closecb().
[2020-04-20 14:09:37] [  LOG]     raop: Send error for 'Kitchen': Connection refused
[2020-04-20 14:09:37] [DEBUG]     raop: Cleaning up failed session (deferred) on device 'Kitchen'
[2020-04-20 14:09:37] [DEBUG]   player: Callback request received, id is 0
[2020-04-20 14:09:37] [DEBUG]   player: Making deferred callback to device_streaming_cb, id was 0
[2020-04-20 14:09:37] [DEBUG]   player: Callback from AirPlay to device_streaming_cb (status -1)
[2020-04-20 14:09:37] [  LOG]   player: The AirPlay device 'Kitchen' FAILED
[2020-04-20 14:09:37] [DEBUG]   player: 1. Active callback: device_streaming_cb
**[2020-04-20 14:09:39] [DEBUG]     mdns: Avahi Resolver: resolved service 'D4909CED1632@Kitchen' type '_raop._tcp' proto 0, host Kitchen.local, address 192.168.0.44
[2020-04-20 14:09:39] [DEBUG]     mdns: Connection test to 192.168.0.44:7000 completed successfully**
[2020-04-20 14:09:39] [DEBUG]     raop: Event for AirPlay device 'Kitchen' (port 7000, id d4909ced1632)
[2020-04-20 14:09:39] [ INFO]     raop: Adding AirPlay device 'Kitchen': password: 0, verification: 0, encrypt: 0, authsetup: 0, metadata: 1, type Other, address 192.168.0.44:7000
[2020-04-20 14:09:39] [DEBUG]      mpd: Asynchronous listener callback called with event type 8.
[2020-04-20 14:09:39] [DEBUG]      mpd: Notify clients waiting for idle results: 8
[2020-04-20 14:09:40] [DEBUG]      web: notify callback reason: 11
[2020-04-20 14:09:40] [DEBUG]      web: JSON api request: '/api/outputs'
[2020-04-20 14:10:07] [DEBUG]     raop: keep_alive: Sending OPTIONS to 'Family Room'
[2020-04-20 14:10:07] [  LOG]     raop: ***** !rs->reqs_in_flight case: not calling evtrtsp_connection_set_closecb().

**Comment by ejurgensen** [comment]
Looks like you are not running a build from the branch? The new branch does not use OPTIONS for keep_alive, so you shouldn't see "keep_alive: Sending OPTIONS"

**Comment by mixographer** [comment]
I did checkout the new branch, and re-built. 

git status shows:
On branch atv_disconnect3

I'll check what I'm doing wrong here

**Comment by saury** [comment]
@ejurgensen 
Is there any instruction for building a docker image for this project?
BTW, is this [repo](https://hub.docker.com/r/linuxserver/daapd) your official image in docker hub?

**Comment by ejurgensen** [comment]
When you have the right version running you should see messages like the below in the log every 30 sec.:

```
[DEBUG]     raop: Metadata sr=44100, pos_ms=0, len_ms=2667441, start=3264587167, display=3264556447, pos=3264587167, end=3382221315, rtptime=3264587167
[DEBUG]     raop: send_progress: Sending SET_PARAMETER to 'ATV4'
[DEBUG]     raop: Metadata sr=44100, pos_ms=0, len_ms=2667441, start=3265910335, display=3265879615, pos=3265910335, end=3383544483, rtptime=3265910335
[DEBUG]     raop: send_progress: Sending SET_PARAMETER to 'ATV4'
```

@saury I don't maintain a docker repo, and please keep the thread on topic

**Comment by mixographer** [comment]
ok. Thanks. I cleaned up my repository and checked-out atv_disconnect3 again. 

My homepod is working well, but my ATV 'Family Room' dropped after some time. I'm playing 7:00 songs. 

[2020-04-21 10:16:12] [DEBUG]     raop: Metadata sr=44100, pos_ms=0, len_ms=439507, start=1915660690, display=1915629970, pos=1915660690, end=1935042948, rtptime=1915660690
[2020-04-21 10:16:12] [DEBUG]     raop: send_progress: Sending SET_PARAMETER to 'Family Room'
[2020-04-21 10:16:12] [DEBUG]     raop: Metadata sr=44100, pos_ms=0, len_ms=439507, start=1915660690, display=1915629970, pos=1915660690, end=1935042948, rtptime=1915660690
[2020-04-21 10:16:12] [DEBUG]     raop: send_progress: Sending SET_PARAMETER to 'Kitchen'
[2020-04-21 10:16:12] [ WARN]     raop: SET_PARAMETER metadata/artwork/progress request to 'Family Room' failed (proceeding anyway): 0 (null)
[2020-04-21 10:16:12] [  LOG]     raop: Device 'Family Room' closed RTSP connection
[2020-04-21 10:16:12] [DEBUG]     raop: Cleaning up failed session (deferred) on device 'Family Room'
[2020-04-21 10:16:12] [DEBUG]   player: Callback request received, id is 1
[2020-04-21 10:16:12] [DEBUG]   player: Making deferred callback to device_streaming_cb, id was 1
[2020-04-21 10:16:12] [DEBUG]   player: Callback from AirPlay to device_streaming_cb (status -1)
[2020-04-21 10:16:12] [  LOG]   player: The AirPlay device 'Family Room' FAILED
[2020-04-21 10:16:12] [DEBUG]   player: 0. Active callback: device_streaming_cb
[2020-04-21 10:16:14] [DEBUG]     mdns: Avahi Resolver: resolved service 'C869CD343C6A@Family Room' type '_raop._tcp' proto 0, host Family-Room.local, address 192.168.0.38
[2020-04-21 10:16:14] [DEBUG]     mdns: Connection test to 192.168.0.38:7000 completed successfully
[2020-04-21 10:16:14] [DEBUG]     raop: Event for AirPlay device 'Family Room' (port 7000, id c869cd343c6a)
[2020-04-21 10:16:14] [ INFO]     raop: Adding AirPlay device 'Family Room': password: 0, verification: 1, encrypt: 0, authsetup: 0, metadata: 1, type AppleTV4, address 192.168.0.38:7000
[2020-04-21 10:16:14] [DEBUG]      mpd: Asynchronous listener callback called with event type 8.
[2020-04-21 10:16:14] [DEBUG]      mpd: Notify clients waiting for idle results: 8
[2020-04-21 10:16:15] [DEBUG]      web: notify callback reason: 11
[2020-04-21 10:16:15] [DEBUG]      web: JSON api request: '/api/outputs'
[2020-04-21 10:16:42] [DEBUG]     raop: Metadata sr=44100, pos_ms=0, len_ms=439507, start=1916983506, display=1916952786, pos=1916983506, end=1936365764, rtptime=1916983506
[2020-04-21 10:16:42] [DEBUG]     raop: send_progress: Sending SET_PARAMETER to 'Kitchen'


I can reconnect immediately to the ATV, so the failure is short-lived. The ATV also is requesting retransmit very often and the HomePod is not. 

[2020-04-21 10:22:14] [DEBUG]     raop: Got retransmit request from 'Family Room': seqnum 22467 (len 1), last RTP session seqnum 22476 (len 1000)
[2020-04-21 10:22:21] [DEBUG]     raop: Got retransmit request from 'Family Room': seqnum 23250 (len 1), last RTP session seqnum 23264 (len 1000)
[2020-04-21 10:22:21] [DEBUG]     raop: Got retransmit request from 'Family Room': seqnum 23363 (len 1), last RTP session seqnum 23367 (len 1000)
[2020-04-21 10:22:23] [DEBUG]     raop: Got retransmit request from 'Family Room': seqnum 23514 (len 1), last RTP session seqnum 23527 (len 1000)
[2020-04-21 10:22:23] [DEBUG]     raop: Got retransmit request from 'Family Room': seqnum 23583 (len 1), last RTP session seqnum 23595 (len 1000)
[2020-04-21 10:22:27] [DEBUG]     raop: Got retransmit request from 'Family Room': seqnum 24027 (len 1), last RTP session seqnum 24036 (len 1000)
[2020-04-21 10:22:27] [DEBUG]     raop: Got retransmit request from 'Family Room': seqnum 24039 (len 1), last RTP session seqnum 24052 (len 1000)
[2020-04-21 10:22:27] [DEBUG]     raop: Got retransmit request from 'Family Room': seqnum 24040 (len 1), last RTP session seqnum 24052 (len 1000)
[2020-04-21 10:22:28] [DEBUG]     raop: Got retransmit request from 'Family Room': seqnum 24153 (len 1), last RTP session seqnum 24166 (len 1000)

**Comment by ejurgensen** [comment]
Thanks for testing. If the Homepod works I think there is still hope for this fix. Let's try again when I have fixed the metadata content.

**Comment by saury** [comment]
@ejurgensen Sorry for interrupting, I was intend to help testing this branch and not sure how to build by cloning this repo. This homepod disconnecting issue also annoying me for a long time. I'm using that docker image to setup the streaming service and the link of its web interface leads me to here, so I thought that might be the official image.

**Comment by holly79** [comment]
@ejurgensen: I compiled the new version and the tests are on going. I'll let you know the result asap.

@saury: you can use this [link](https://github.com/ejurgensen/forked-daapd/blob/master/INSTALL) for instruction to build your own version. I installed a version with apt and build a new version like instruction. Google for checkout a specific  branche.

**Comment by ejurgensen** [comment]
I have added some commits to the [atv_disconnect3](https://github.com/ejurgensen/forked-daapd/tree/atv_disconnect3) branch that I think fix the metadata content and a few other issues I saw while testing. Still needs more testing, though. I still haven't seen any disconnects from my own ATV4 with these changes, so that's something. I did notice an issue with ipv6, which I need to look into. So right now I recommend testing with ipv6 disabled in the config file.

**Comment by raufis27** [comment]
Atv_disconnect3:
testing on 2 HomePods. One of them failed minute ago:

`[2020-04-22 22:26:20] [DEBUG]   player: Player status: playing
[2020-04-22 22:26:21] [DEBUG]     raop: pos_ms=1020, len_ms=0, startup=0, metadata.pts=799351.603860207, player.ts=799558.583860207, diff_ms=206980, elapsed_ms=208000
[2020-04-22 22:26:21] [DEBUG]     raop: start=1477821918, display=1477791198, pos=1486994718, end=1486994718, rtp_session.pos=1487083038, cur_stamp.pos=1486994718
[2020-04-22 22:26:21] [DEBUG]     raop: send_progress: Sending SET_PARAMETER to 'iForest'
[2020-04-22 22:26:21] [DEBUG]     raop: pos_ms=1020, len_ms=0, startup=0, metadata.pts=799351.603860207, player.ts=799558.583860207, diff_ms=206980, elapsed_ms=208000
[2020-04-22 22:26:21] [DEBUG]     raop: start=1477821918, display=1477791198, pos=1486994718, end=1486994718, rtp_session.pos=1487083038, cur_stamp.pos=1486994718
[2020-04-22 22:26:21] [DEBUG]     raop: send_progress: Sending SET_PARAMETER to 'LR-White'
[2020-04-22 22:26:21] [DEBUG]     raop: pos_ms=1020, len_ms=0, startup=0, metadata.pts=799351.603860207, player.ts=799558.583860207, diff_ms=206980, elapsed_ms=208000
[2020-04-22 22:26:21] [DEBUG]     raop: start=1477821918, display=1477791198, pos=1486994718, end=1486994718, rtp_session.pos=1487083038, cur_stamp.pos=1486994718
[2020-04-22 22:26:21] [DEBUG]     raop: send_progress: Sending SET_PARAMETER to 'LR-Black'
[2020-04-22 22:26:21] [DEBUG]     raop: pos_ms=1020, len_ms=0, startup=0, metadata.pts=799351.603860207, player.ts=799558.583860207, diff_ms=206980, elapsed_ms=208000
[2020-04-22 22:26:21] [DEBUG]     raop: start=1477821918, display=1477791198, pos=1486994718, end=1486994718, rtp_session.pos=1487083038, cur_stamp.pos=1486994718
[2020-04-22 22:26:21] [DEBUG]     raop: send_progress: Sending SET_PARAMETER to 'BasementATV'
[2020-04-22 22:26:21] [ WARN]     raop: SET_PARAMETER metadata/artwork/progress request to 'LR-White' failed (proceeding anyway): 0 (null)
[2020-04-22 22:26:21] [  LOG]     raop: Device 'LR-White' closed RTSP connection
[2020-04-22 22:26:21] [DEBUG]     raop: Cleaning up failed session (deferred) on device 'LR-White'
[2020-04-22 22:26:21] [DEBUG]   player: Callback request received, id is 2
[2020-04-22 22:26:21] [DEBUG]   player: Making deferred callback to device_streaming_cb, id was 2
[2020-04-22 22:26:21] [DEBUG]   player: Callback from AirPlay to device_streaming_cb (status -1)
[2020-04-22 22:26:21] [  LOG]   player: The AirPlay device 'LR-White' FAILED
[2020-04-22 22:26:21] [DEBUG]   player: 0. Active callback: device_streaming_cb
[2020-04-22 22:26:21] [DEBUG]   player: 1. Active callback: device_streaming_cb
[2020-04-22 22:26:21] [DEBUG]   player: 3. Active callback: device_streaming_cb
[2020-04-22 22:26:22] [DEBUG]     mdns: Avahi Resolver: resolved service 'D4909CDF6D6F@LR-White' type '_raop._tcp' proto 0, host LR-White.local, address 192.168.200.160
[2020-04-22 22:26:22] [DEBUG]     mdns: Connection test to 10.10.100.160:7000 completed successfully
[2020-04-22 22:26:22] [DEBUG]     raop: Event for AirPlay device 'LR-White' (port 7000, id d4909cdf6d6f)
[2020-04-22 22:26:22] [ INFO]     raop: Adding AirPlay device 'LR-White': password: 0, verification: 0, encrypt: 0, authsetup: 0, metadata: 7, type Other, address 10.10.100.160:7000`

**Comment by raufis27** [comment]
Update: 
As you see from previous log, I have 2 HomePods. One of them LRWhite disconnected ~2 min after I started playing internet radio. LRBlack didn't. Then I enabled LRWhite again and continued play. So far (2 hours 15 min) everything is stable. Both playing fine. I'll continue testing.

**Comment by ejurgensen** [comment]
Really weird behavior... what sort of error is so sporadic?

I suppose we can conclude that the changes don't fix the issue, but are they an improvement? I.e. should I include them? I'm not sure how frequent the disconnects were before. I hope everyone who is testing can pitch in on that.

**Comment by saury** [comment]
I tested it yesterday by playing m3u streaming and it never disconnect when playing on one homepod for 2 hours. But one of my two homepods disconnected after about 10 minutes by setting outputs to 2 homepods. The only problem is that it seldom pause for several seconds during playing.
It seems this fix version works and is much more stable than the version on master branch.

**Comment by raufis27** [comment]
I agree that this version is more stable. Both HomePods playing for past hour without disconnect. I'll keep testing.

**Comment by ejurgensen** [comment]
Thanks for the feedback. My own testing also indicates that it works better now, and have also confirmed that the metadata content is correct. So I have merged the changes into master now.

I will close the issue, even though there are still sporadic disconnects. It is unlikely that I can fix those, but if someone is able to find a pattern or even better reproducebility in these disconnects, then please open a new issue.

**Comment by NorbertRoller** [comment]
Will you add the changes of the atv_disconnect3 to the main branch or can you build a raspberry version for me to try ?

**Comment by skabber** [comment]
I’m going to keep poking around this. I just had a HomePod drop after a 15+ minute session.
`[2020-04-27 08:45:53] [  LOG]     raop: Device 'Kitchen' closed RTSP connection
[2020-04-27 08:45:54] [  LOG]   player: The AirPlay device 'Kitchen' FAILED
[2020-04-27 08:45:54] [  LOG]   player: The AirPlay device 'Living Room' FAILED `

**Comment by raufis27** [comment]
I have same problem now, it disconnects randomly. It will be good have an option to auto enable speakers.

**Comment by ejurgensen** [comment]
Depressing that the issue isn't completely gone. Hope someone finds a clue.

@NorbertRoller the changes are already merged into the main branch. Here are test builds for RPi you can try. When I make the next actual RPi release apt should automatically replace them.
Stretch: http://www.gyfgafguf.dk/temp/forked-daapd_27.1.105~rc2-2_armhf.deb
Buster: http://www.gyfgafguf.dk/temp/forked-daapd_27.1.105~rc2-3_armhf.deb

**Comment by ejurgensen** [comment]
Yes, auto enabling would help. It is not completely trivial, one of the reasons is that usually reconnecting immediately won't work, so forked-daapd would have to wait a bit.

**Comment by mixographer** [comment]
According to my log files, it's usually less than 2 seconds. I agree it would be nice to watch for the disconnect and then auto reconnect. I'd rather have a short dropout than have to reach for the phone to manually reconnect. 

I wonder if this is a problem in the underlying mDNS lookups in the system, rather than with forked-daapd in particular. Thanks for spending so much effort on looking at this one.

**Comment by holly79** [comment]
I tested it with different kind of sources. Local mp3, Internet streams and Spotify. Only local mp3 with 0:27 sec running over hours (in a loop). Mp3 with 0:53 or 2:32 (in a loop) stopped on one or more Homepods after few minutes. Internet Streams and Spotify stopped on one or more Homepods after few minutes, too. After I changed the keep_alive_tv from 30 sec to 15 sec, it was running Internet streams over hours in the night. I'm trying 25 sec and it seems it run much more stable for Internet streams and Spotify as 30 sec. Test is not finished! I'll let you know more in one or two days.

**Comment by ejurgensen** [comment]
Thanks for the thorough test. Could you also test if setting the "true" [in this line](https://github.com/ejurgensen/forked-daapd/blob/master/src/outputs/raop.c#L2458) to "false" makes a difference? It determines how much metadata gets sent to the speaker. "true" means it only sends progress, while "false" means it sends everything.

I am pretty sure the type of input doesn't make a difference (it's all the same to the speaker), only how long the track is, since that affects how often metadata is sent.

I am working on adding the ability for forked-daapd to auto-reconnect. It should be ready for testing in a few days.

**Comment by mynameisdaniel32** [comment]
Just on my experience with this - I was having the same problems with random disconnects regularly from 5 minutes to (at best) 2 hours when streaming internet radio to two HomePods. Updating last week and now both HomePods are streaming perfectly for 8+ hours almost every day since then. Not a single unexpected dropout since. Thanks!

I'm wondering if some config I did a while ago to troubleshoot is helping me here, might be worth trying for these more recent dropouts?

```
airplay "Bedroom HomePod" {
	permanent = true
}

airplay "Living Room HomePod" {
	permanent = true
}
```

**Comment by holly79** [comment]
@mynameisdaniel32: This config is only to solve the disappearing from outputs list. It doesn't help to solve the disconnect issue.

@ejurgensen: I test the code with keep alive of 30 sec and false in L2458. It was the same behaviour as before. A keep a live of 25 sec and false in L2458 is working like 25 sec and true in L2458. From my perspective, it makes no difference. 

In summary, the 25 seconds are sufficient to achieve stable streaming.

**Comment by NorbertRoller** [comment]
If you share a build version for the Raspberry, I can contribute to the testing.

PS: I covered the issue with my homeserver very similarily. Every Minute it switches it back on. But a minute will show dropouts of something between 10 sec - 50 sec.

**Comment by ejurgensen** [comment]
I have now added commits to master so that:

1) forked-daapd automatically reconnects 5 second after a speaker fails (you will see "The AirPlay device 'XXX' failed - attempting reconnect in 5 sec")
2) the interval is 25 sec instead of 30 sec

I also had some other refactoring/improvements/fixes related to speaker selection that I had on my todo-list, so that is also bundled into these changes.

Here are builds for testing on the RPi:
Stretch: http://www.gyfgafguf.dk/temp/forked-daapd_27.1.105~rc4-2_armhf.deb
Buster: http://www.gyfgafguf.dk/temp/forked-daapd_27.1.105~rc4-3_armhf.deb

**Comment by holly79** [comment]
The new build is working fine in my environment.

**Comment by ejurgensen** [comment]
Closing this again, since it is now at least somewhat stable + can now reconnect on error. I've also updated my RPi builds to include the fix.
